src/lowlevel/pin_util.rs

```diff
@@ -0,0 +1,87 @@
+use std::ops::Deref;
+use std::pin::Pin;
+use std::sync::Arc;
+
+use tokio::sync::{Mutex as MutexAsync, MutexGuard as MutexAsyncGuard};
+
+#[derive(Debug)]
+pub(super) struct PinnedArc<T>(Arc<T>);
```

Comment:
I think I'm overall confused about why you have this type. Why is this not just `Pin<Arc<T>>`, which is supported by the standard library (e.g., [`Arc::pin`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.pin))? If that API is insufficient in some way, that feels like we should file an issue with rust-lang/rust stating what the problem is. I _think_ everything you're doing here is possible to do with just `Arc::pin`, `Pin<Arc<T>>`, and `Pin::as_ref`.

Reply:
It's because of the Arc::get_mut, which is used only in WriteEnd::new to send the initial message.

I agree that this is not a good design and will replace that with acquiring the mutex instead.

Reply:
I have refactored the code, removed one public API of `SharedData` which does not provide any value and removed use of `PinnedArc`.

However, I cannot use `Arc::strong_count` with `Pin<Arc>`, so I ended up adding a new function `pinned_arc_strong_count` to this file.

Reply:
If we can obtain `&P` from `&Pin<P>`, then I wouldn't have to write this `unsafe` function.

Obtaining `&P` from `&Pin<P>` should be safe, as it cannot move into/from the pinned object anyway.

@jonhoo Do you think it is a good idea to add file an issue for this new function?

Reply:
Also, the `PinnedMutexAsync` is also a pain.

Perhaps it is a good idea to open a PR in [tokio](github.com/tokio-rs/tokio)?

Reply:
Huh, interesting, getting a `&P` from `&Pin<P>` _should_ be safe, I agree, which means it should arguably be in the standard library. I found [this discussion thread](https://internals.rust-lang.org/t/arc-and-pin/16047) talking about this though, and that brought up [this crate](https://docs.rs/pinnable/0.1.1/pinnable/struct.Mutex.html) which you might find interesting!

Reply:
[`pinnable::Mutex`](https://docs.rs/pinnable/0.1.1/pinnable/struct.Mutex.html) is nice but we need its async counterpart.

About getting `&P` from `&Pin<P>` into std, how to request for new function in std?
From what I read, it seems that I need to create a new RFC to explain the motivation for this.

Reply:
I would just open a PR for that one. It's not a major language change, so shouldn't need an RFC. You'd add it somewhere near here: https://github.com/rust-lang/rust/blob/c7b0452ece11bf714f7cf2003747231931504d59/library/core/src/pin.rs#L591

Reply:
@jonhoo I have created a fork [here](https://github.com/NobodyXu/rust/tree/feature/pin) and added the new function according to your suggestion.

Since I am really don't know how to modify the std library properly and I am not sure about the name of the new function, I really would like your feedback before submitting it as a PR.

If you think it is more appropriate to just open a PR and leave your feedback, just tell me and I will do that.

Reply:
> I would just open a PR for that one. It's not a major language change, so shouldn't need an RFC. You'd add it somewhere near here: https://github.com/rust-lang/rust/blob/c7b0452ece11bf714f7cf2003747231931504d59/library/core/src/pin.rs#L591

@jonhoo I've opened a PR https://github.com/rust-lang/rust/pull/98325

Reply:
Sorry, I completely missed this notification! I think opening a PR was the right way to go — chances are the libs team will be able to give you better guidance than I'd be able to anyway :)

Reply:
Thank you for taking your time to review this PR and give me advice on opening a PR on rust-lang/rust !